hotsos/core/plugins/juju/resources.py

```diff
@@ -204,3 +204,16 @@ def charm_names(self):
             return []
 
         return list(self.charms.keys())
+
+
+class JujuBinaryInterface(JujuBase):
```

Comment:
Do we need to have such a BinaryInterface class for each snap e.g. `kubectl`, `lxd`, etc?

Reply:
No because snaps are covered by YRequirementTypeSnap

Reply:
But we do for `juju` only because older ones are apt pkg?

Reply:
juju is installed in different ways depending on the host, the juju client is typically a snap but the juju controller is not installed by a (apt/snap) package so we have to query the installed binary directly instead. Here is an example from a juju controller:

```
root@juju-da71e1-controller-0:/home/ubuntu# dpkg -l| grep juju
root@juju-da71e1-controller-0:/home/ubuntu# snap list| grep juju
juju-db  4.4.18         160    4.4/stable     juju-qa      -
root@juju-da71e1-controller-0:/home/ubuntu# dpkg -S /etc/systemd/system/jujud-machine-0.service
dpkg-query: no path found matching pattern /etc/systemd/system/jujud-machine-0.service
root@juju-da71e1-controller-0:/home/ubuntu# dpkg -S /var/lib/juju/tools/machine-0/jujud
dpkg-query: no path found matching pattern /var/lib/juju/tools/machine-0/jujud
root@juju-da71e1-controller-0:/home/ubuntu# /var/lib/juju/tools/machine-0/jujud --version
2.9.49-ubuntu-amd64
```

Reply:
to be fair, in this patch the JujuBinaryInterface.get_version() is calling JujuMachine.version which gets the version from the config file. I think I will extend JujuMachine to first actually query the binary then go to config as a fallback.

Reply:
Thanks for the details.